os/arch/arm/src/amebad/amebad_flash.c

```diff
@@ -156,59 +154,39 @@ static int amebad_erase(FAR struct mtd_dev_s *dev, off_t startblock, size_t nblo
 
 static ssize_t amebad_flash_write(size_t addr, const void *buf, size_t length)
 {
-    flash_t flash;
     int32_t result = 0;
-    ssize_t ret = 0;
     irqstate_t irqs;
 
     /* Disable IRQs while erasing sector */
     irqs = irqsave();
-    if ((addr & 0xff) != 0) {
-        //! value should be 256 byte alignment
-        ret = -EFAULT;
-    } else if ((length & 0x3) != 0) {
-        ret = -EINVAL;
-    } else {
-        char *adpt_buf = (char *)kmm_malloc(length);
-        if (adpt_buf == NULL) {
-            ret = -EPERM;
-        } else {
-            memcpy(adpt_buf, buf, length);
-            result = flash_stream_write(&flash, addr, length, adpt_buf);
 
-            if (result < 0) {
-                ret = -EIO;
-            } else {
-                ret = length;
-            }
-            kmm_free(adpt_buf);
-            adpt_buf = NULL;
-        }
-    }
+    result = flash_stream_write(NULL, addr, length, buf);
 
     /* Restore IRQs */
     irqrestore(irqs);
-    return ret;
+
+    if (result < 0) {

```

Comment:
```
if (result < 0) {
   return -EIO;
}
return result;
```
But flash_stream_write should return error value. for example
```
int  flash_stream_write(flash_t *obj, u32 address, u32 len, u8 * data)
{
    if (data == NULL) {
        return -EINVAL;
    }
    if (addr is_out_of_valid_address) {
        return -EFAULT;
     }
    ...and something more
}
```

Reply:
This error return you suggested will be done in amebad_flash.c instead of flash_api.c .
And it will be updated with other changes, thanks!